File: cli/src/lib/global-reconciliation.ts
import {
  findContextAnchorCandidates,
  getAnchorContextScope,
  type AnchorContextIndex,
  type ContextAnchorCandidate,
} from "./anchor-context.js";
import type { Comment, ReanchorResult } from "./types.js";

const LOCAL_LANDMARK_WINDOW = 30;
const MIN_LANDMARK_SUPPORT = 0.65;
const MIN_SUPPORTING_LANDMARKS = 2;
const MIN_GLOBAL_MARGIN = 0.08;
const MAX_RECONCILIATION_ROUNDS = 4;

interface Landmark {
  sourceLine: number;
  targetLine: number;
  scope: string;
}

interface ScoredCandidate {
  candidate: ContextAnchorCandidate;
  globalScore: number;
  support: number;
  supportingLandmarks: number;
}

/**
 * Resolve uncertain comments from nearby high-confidence anchors.
 *
 * Each landmark votes for a local source-to-target displacement rather than
 * global document order, allowing whole sections to move independently.
 * Newly confirmed exact anchors become landmarks in subsequent rounds.
 */
export function reconcileCommentAnchors(
  comments: Comment[],
  results: ReanchorResult[],
  anchorContext: AnchorContextIndex,
): ReanchorResult[] {
  const reconciled = results.map((result) => ({ ...result }));
  if (
    comments.length < MIN_SUPPORTING_LANDMARKS + 1
    || !reconciled.some((result) => result.status === "ambiguous")
  ) {
    return reconciled;
  }
  const resultIndexes = new Map(
    reconciled.map((result, index) => [result.commentId, index]),
  );
  const landmarks = collectLandmarks(comments, reconciled, anchorContext);
  const landmarkCounts = new Map<string, number>();
  for (const landmark of landmarks) {
    landmarkCounts.set(
      landmark.scope,
      (landmarkCounts.get(landmark.scope) ?? 0) + 1,
    );
  }

  for (let round = 0; round < MAX_RECONCILIATION_ROUNDS; round += 1) {
    let changed = false;
    for (const comment of comments) {
      if (comment.line == null || !comment.selected_text) continue;
      const scope = getAnchorContextScope(comment, anchorContext);
      if (scope == null) continue;
      if ((landmarkCounts.get(scope) ?? 0) < MIN_SUPPORTING_LANDMARKS) {
        continue;
      }
      const resultIndex = resultIndexes.get(comment.id);
      if (resultIndex == null) continue;
      const result = reconciled[resultIndex];
      if (result.status !== "ambiguous") continue;

      const candidates = deduplicateCandidates(
        findContextAnchorCandidates(comment, anchorContext),
      ).slice(0, 8);
      if (candidates.length < 2) continue;
      const ranked = candidates
        .map((candidate) =>
          scoreWithLandmarks(
            comment.line as number,
            scope,
            candidate,
            landmarks,
          )
        )
        .sort((left, right) =>
          right.globalScore - left.globalScore
          || right.candidate.score - left.candidate.score
          || left.candidate.line - right.candidate.line
        );
      const best = ranked[0];
      const second = ranked[1];
      if (
        !best.candidate.exact
        || best.supportingLandmarks < MIN_SUPPORTING_LANDMARKS
        || best.support < MIN_LANDMARK_SUPPORT
        || best.globalScore - second.globalScore < MIN_GLOBAL_MARGIN
      ) {
        continue;
      }

      reconciled[resultIndex] = {
        commentId: comment.id,
        status: "anchored",
        score: Math.min(0.99, 0.8 + best.support * 0.19),
        newLine: best.candidate.line,
        newEndLine: best.candidate.endLine,
        newStartColumn: best.candidate.startColumn,
        newEndColumn: best.candidate.endColumn,
        reason:
          `Global reconciliation selected this candidate from `
          + `${best.supportingLandmarks} nearby landmark(s) `
          + `(support ${best.support.toFixed(3)}, margin `
          + `${(best.globalScore - second.globalScore).toFixed(3)}).`,
      };
      changed = true;

      landmarks.push({
        sourceLine: comment.line,
        targetLine: best.candidate.line,
        scope,
      });
      landmarkCounts.set(scope, (landmarkCounts.get(scope) ?? 0) + 1);
    }
    if (!changed) break;
  }

  return reconciled;
}

function collectLandmarks(
  comments: Comment[],
  results: ReanchorResult[],
  anchorContext: AnchorContextIndex,
): Landmark[] {
  const resultMap = new Map(results.map((result) => [result.commentId, result]));
  return comments.flatMap((comment): Landmark[] => {
    const result = resultMap.get(comment.id);
    if (
      comment.line == null
      || result?.newLine == null
      || (result.status !== "anchored" && result.status !== "shifted")
      || result.score < 0.99
    ) {
      return [];
    }
    const scope = getAnchorContextScope(comment, anchorContext);
    return scope == null
      ? []
      : [{ sourceLine: comment.line, targetLine: result.newLine, scope }];
  });
}

function scoreWithLandmarks(
  sourceLine: number,
  scope: string,
  candidate: ContextAnchorCandidate,
  landmarks: Landmark[],
): ScoredCandidate {
  let weightedSupport = 0;
  let totalWeight = 0;
  let supportingLandmarks = 0;
  const candidateShift = candidate.line - sourceLine;

  for (const landmark of landmarks) {
    if (landmark.scope !== scope) continue;
    const sourceDistance = Math.abs(sourceLine - landmark.sourceLine);
    if (sourceDistance === 0 || sourceDistance > LOCAL_LANDMARK_WINDOW) continue;
    const landmarkShift = landmark.targetLine - landmark.sourceLine;
    const shiftError = Math.abs(candidateShift - landmarkShift);
    const agreement = Math.max(0, 1 - shiftError / 8);
    const weight = 1 / (1 + sourceDistance / 4);
    weightedSupport += agreement * weight;
    totalWeight += weight;
    if (agreement >= MIN_LANDMARK_SUPPORT) supportingLandmarks += 1;
  }

  const support = totalWeight > 0 ? weightedSupport / totalWeight : 0;
  return {
    candidate,
    globalScore:
      candidate.score * 0.5 + support * 0.45 + (candidate.exact ? 0.05 : 0),
    support,
    supportingLandmarks,
  };
}

function deduplicateCandidates(
  candidates: ContextAnchorCandidate[],
): ContextAnchorCandidate[] {
  const unique = new Map<string, ContextAnchorCandidate>();
  for (const candidate of candidates) {
    const key = [
      candidate.line,
      candidate.endLine,
      candidate.startColumn ?? "",
      candidate.endColumn ?? "",
    ].join(":");
    const existing = unique.get(key);
    if (!existing || candidate.score > existing.score) {
      unique.set(key, candidate);
    }
  }
  return [...unique.values()].sort((left, right) =>
    right.score - left.score || left.line - right.line
  );
}

File: python/src/mrsf/global_reconciliation.py
"""Heading-scoped multi-comment anchor reconciliation."""

from __future__ import annotations

from dataclasses import dataclass, replace

from .anchor_context import (
    AnchorContextIndex,
    ContextAnchorCandidate,
    find_context_anchor_candidates,
    get_anchor_context_scope,
)
from .types import Comment, ReanchorResult

LOCAL_LANDMARK_WINDOW = 30
MIN_LANDMARK_SUPPORT = 0.65
MIN_SUPPORTING_LANDMARKS = 2
MIN_GLOBAL_MARGIN = 0.08
MAX_RECONCILIATION_ROUNDS = 4


@dataclass
class _Landmark:
    source_line: int
    target_line: int
    scope: str


@dataclass
class _ScoredCandidate:
    candidate: ContextAnchorCandidate
    global_score: float
    support: float
    supporting_landmarks: int


def reconcile_comment_anchors(
    comments: list[Comment],
    results: list[ReanchorResult],
    anchor_context: AnchorContextIndex,
) -> list[ReanchorResult]:
    reconciled = [replace(result) for result in results]
    if len(comments) < 3 or not any(result.status == "ambiguous" for result in reconciled):
        return reconciled
    result_indexes = {result.comment_id: index for index, result in enumerate(reconciled)}
    landmarks = _collect_landmarks(comments, reconciled, anchor_context)
    counts: dict[str, int] = {}
    for landmark in landmarks:
        counts[landmark.scope] = counts.get(landmark.scope, 0) + 1

    for _ in range(MAX_RECONCILIATION_ROUNDS):
        changed = False
        for comment in comments:
            if comment.line is None or not comment.selected_text:
                continue
            scope = get_anchor_context_scope(comment, anchor_context)
            if scope is None or counts.get(scope, 0) < MIN_SUPPORTING_LANDMARKS:
                continue
            result_index = result_indexes.get(comment.id)
            if result_index is None or reconciled[result_index].status != "ambiguous":
                continue
            candidates = _deduplicate(
                find_context_anchor_candidates(comment, anchor_context)
            )[:8]
            if len(candidates) < 2:
                continue
            ranked = sorted(
                (
                    _score_with_landmarks(comment.line, scope, candidate, landmarks)
                    for candidate in candidates
                ),
                key=lambda item: (
                    -item.global_score,
                    -item.candidate.score,
                    item.candidate.line,
                ),
            )
            best, second = ranked[0], ranked[1]
            if (
                not best.candidate.exact
                or best.supporting_landmarks < MIN_SUPPORTING_LANDMARKS
                or best.support < MIN_LANDMARK_SUPPORT
                or best.global_score - second.global_score < MIN_GLOBAL_MARGIN
            ):
                continue
            reconciled[result_index] = ReanchorResult(
                comment_id=comment.id,
                status="anchored",
                score=min(0.99, 0.8 + best.support * 0.19),
                new_line=best.candidate.line,
                new_end_line=best.candidate.end_line,
                new_start_column=best.candidate.start_column,
                new_end_column=best.candidate.end_column,
                reason=(
                    f"Global reconciliation selected this candidate from "
                    f"{best.supporting_landmarks} nearby landmark(s) "
                    f"(support {best.support:.3f}, margin "
                    f"{best.global_score - second.global_score:.3f})."
                ),
            )
            landmarks.append(_Landmark(comment.line, best.candidate.line, scope))
            counts[scope] = counts.get(scope, 0) + 1
            changed = True
        if not changed:
            break
    return reconciled


def _collect_landmarks(
    comments: list[Comment],
    results: list[ReanchorResult],
    context: AnchorContextIndex,
) -> list[_Landmark]:
    result_map = {result.comment_id: result for result in results}
    landmarks: list[_Landmark] = []
    for comment in comments:
        result = result_map.get(comment.id)
        if (
            comment.line is None
            or result is None
            or result.new_line is None
            or result.status not in ("anchored", "shifted")
            or result.score < 0.99
        ):
            continue
        scope = get_anchor_context_scope(comment, context)
        if scope is not None:
            landmarks.append(_Landmark(comment.line, result.new_line, scope))
    return landmarks


def _score_with_landmarks(
    source_line: int,
    scope: str,
    candidate: ContextAnchorCandidate,
    landmarks: list[_Landmark],
) -> _ScoredCandidate:
    weighted = total = 0.0
    supporting = 0
    candidate_shift = candidate.line - source_line
    for landmark in landmarks:
        distance = abs(source_line - landmark.source_line)
        if landmark.scope != scope or distance == 0 or distance > LOCAL_LANDMARK_WINDOW:
            continue
        error = abs(candidate_shift - (landmark.target_line - landmark.source_line))
        agreement = max(0.0, 1 - error / 8)
        weight = 1 / (1 + distance / 4)
        weighted += agreement * weight
        total += weight
        if agreement >= MIN_LANDMARK_SUPPORT:
            supporting += 1
    support = weighted / total if total else 0.0
    score = candidate.score * 0.5 + support * 0.45 + (0.05 if candidate.exact else 0)
    return _ScoredCandidate(candidate, score, support, supporting)


def _deduplicate(candidates: list[ContextAnchorCandidate]) -> list[ContextAnchorCandidate]:
    unique: dict[tuple[int, int, int | None, int | None], ContextAnchorCandidate] = {}
    for candidate in candidates:
        key = (
            candidate.line,
            candidate.end_line,
            candidate.start_column,
            candidate.end_column,
        )
        if key not in unique or candidate.score > unique[key].score:
            unique[key] = candidate
    return sorted(unique.values(), key=lambda item: (-item.score, item.line))
